### ISSUE TITLE
feat(image): text-to-image generation via mflux on /v1/images/generations (Qwen-Image / Qwen-Image-2.1)

### ISSUE THREAD

**Issue by dpalmqvist** [motsognirr/olmlx#723]
Add text-to-image generation via mflux, served on `/v1/images/generations`. Target models are the MLX ports of Qwen-Image (20B MMDiT) and Qwen-Image-2.1 (7.1B DiT + ~8B Qwen3-VL text encoder).

olmlx has no image-generation path today: no diffusion backend in `pyproject.toml`, no image router, and nothing serving `/v1/images/generations`. (The `diffusion` hits in the tree are `engine/dflash/` — block-diffusion *text* speculative decoding, unrelated.)

## Feasibility — verified, not assumed

Checked against a throwaway venv installed at this project's **exact locked versions** (mlx 0.32.0, torch 2.13.0, transformers 5.11.0, hf-hub 1.19.0), not a fresh resolver pick.

**1. No dependency conflict.** mflux 0.20.0 declares `mlx<0.33.0,>=0.32.0` on darwin — character-for-character our pin. Every other requirement is satisfied by the current `uv.lock`:

| mflux requires | uv.lock has |
|---|---|
| `huggingface-hub<2.0,>=1.1.6` | 1.19.0 |
| `transformers<6.0,>=5.5.0` | 5.11.0 |
| `numpy<3.0,>=2.0.1` | 2.4.6 |
| `safetensors<1.0,>=0.4.4` | 0.8.0 |
| `protobuf<8.0,>=4.25.0` | 6.33.6 |
| `torch<3.0,>=2.13.0` | 2.13.0 (exactly at floor) |
| `pillow>=12.3.0` | 12.3.0 (exactly at floor) |

**2. First-class library API**, not a CLI wrapper:

```python
QwenImage21(quantize=8, model_path=None).generate_image(
    seed, prompt, num_inference_steps=40, height=1024, width=1024,
    guidance=1.0, negative_prompt=None, image_path=None, image_strength=None,
) -> GeneratedImage
```

**3. Thread-safety posture is friendly.** Both `QwenImage` and `QwenImage21` are `mlx.nn.Module` subclasses whose submodules are **non-underscore** annotated attributes (`vae`, `transformer`, `text_encoder`). So `model.parameters()` traverses the full tree, and `mx.eval(model.parameters())` + `_materialize_module_buffers(model)` apply verbatim. There is no repeat of the `wrap_flash_moe` trap where `wrapped.parameters()` came back empty because the inner model sat under `self._model`.

**4. PIL image is directly reachable.** `GeneratedImage.image` is a `PIL.Image.Image` (`mflux/utils/generated_image.py:54`), so `response_format: "b64_json"` is a direct encode — no temp-file round trip, no static file server, no reuse of `routers/blobs.py`.

**5. Cancellation has a real hook.** `model.callbacks` is a `CallbackRegistry` (`register`, `in_loop_callbacks`), invoked as `ctx.in_loop(t, latents)` on every diffusion step. The denoise loop's `try` catches only `KeyboardInterrupt`, so a custom exception raised from an in-loop callback propagates cleanly out of `generate_image`. That is our `cancel_event` hook, per step — structurally identical to `_chunked_prefill`'s per-sub-chunk cancel check.

**Not verified:** no weights were downloaded, so there has been no real forward pass. Unconfirmed: whether lazy underscore buffers actually exist inside the DiT/VAE (module structure says they would be caught if they do), real memory footprint at `-q 8`, and generation wall-clock.

## Architecture

Image models are the **fourth non-LM modality**, after whisper (#366), TTS (#367), and the reranker (#369). They follow the established pattern exactly: `_detect_model_kind` sniffs the checkpoint, a dedicated loader returns the standard 5-tuple with `tokenizer=None` / no chat template / no speculative decoder, and a flag on `LoadedModel` gates the LLM-only paths.

```
POST /v1/images/generations
  -> routers/images.py
  -> engine/inference.py::generate_image      (mirrors generate_speech, inference.py:5031)
       - manager.ensure_loaded(model, keep_alive, pin=True)
       - reject non-image model -> ValueError -> 400   (mirrors generate_rerank, inference.py:4887)
       - async with _inference_locked(...)
       - asyncio.to_thread(worker)  ->  mflux generate_image
  -> b64_json response
```

### Resolution and download: mflux-resolved (decided)

**mflux resolves and downloads these models; olmlx does not.** `Qwen21Initializer.init` does `path = model_path if model_path else model_config.model_name`, so passing `model_path=None` with a `ModelConfig` hands the whole fetch to mflux, which owns its own registry (`AVAILABLE_MODELS`, 35 entries) and its own `get_download_patterns()`.

Rejected alternative: pre-downloading through our existing snapshot path and passing `model_path`. That requires replicating mflux's per-component download patterns and keeping them in sync across mflux releases, for no benefit — mflux checkpoints use a **diffusers-style layout** (components under `vae/`, `transformer/`, `text_encoder/`, `processor/`, per `Qwen21WeightDefinition.get_components()`) that our snapshot path has no reason to learn.

Accepted consequence: image models do **not** participate in our registry-driven download bookkeeping. Document in `CLAUDE.md`.

### Detection: explicit declaration, never sniffing

Two independent reasons auto-detection cannot work the way it does for the other modalities:

**1. There is no top-level `config.json`.** The diffusers layout has none, so `_detect_model_kind`'s config read returns `"unknown"` for every Qwen-Image repo.

**2. `ModelConfig.from_name()` is a loose substring matcher and silently misclassifies text models.** Verified:

```
Qwen/Qwen-Image-2.1                -> OK   base=Qwen/Qwen-Image-2.1        (correct)
Qwen/Qwen3-32B-4bit                -> OK   base=Qwen/Qwen-Image-2512       (!!! a text LLM)
mlx-community/Qwen-Image-2.1-4bit  -> OK   base=Qwen/Qwen-Image-2.1        (correct)
totally/bogus-repo                 -> ModelConfigError
```

`Qwen/Qwen3-32B-4bit` is a plain text model and `from_name` happily resolves it to a Qwen-Image base. Using `from_name` as the discriminator would route **every `Qwen/Qwen3-*` text model** into the image loader. Do not use it for detection under any circumstances. It is fine *after* a model is already known to be an image model, to turn a declared repo id into a `ModelConfig`.

**Therefore: image models are declared explicitly in `models.json`**, mirroring how `adapters` is already an explicit section rather than something we sniff:

```json
{
  "qwen-image:2.1": { "type": "image", "hf_path": "Qwen/Qwen-Image-2.1" },
  "qwen-image:20b": { "type": "image", "hf_path": "Qwen/Qwen-Image-2512" }
}
```

`_detect_model_kind` returns `"image"` only for entries carrying the explicit marker. As a guard, the loader should additionally validate the declared `hf_path` against the **exact** mflux alias/`model_name` set (116 strings; exact match, never substring) and raise a clear `ValueError` on a miss, so a typo fails at load with a useful message instead of downloading tens of GB of the wrong thing.

Note: mflux's `qwen-image` alias resolves to `Qwen/Qwen-Image-2512` (the Dec 2025 refresh), **not** the original `Qwen/Qwen-Image`.

### Touch points

New files:
- `olmlx/routers/images.py`
- `olmlx/schemas/images.py`
- `tests/test_images_*.py`

Modified:
- `olmlx/engine/registry.py` — parse the explicit `"type": "image"` marker on a model entry and carry it through to the manager
- `olmlx/engine/model_manager.py`
  - `_detect_model_kind` — return `"image"` from the explicit registry marker, short-circuiting **before** the `config.json` download/read (there is no `config.json` to read)
  - `_load_model` dispatch (~:2784) — add `_load_model_image`
  - new `_load_model_image` — mirrors `_load_model_tts` (:2905), including `ImportError` -> `ValueError` so a missing extra surfaces as a 400, not an opaque 500
  - `:1178` — set `is_image=(_model_kind == "image")`
  - `:470` — add `is_image` to the adapter-base rejection reason list
  - `:1875` — add `is_image` to the prompt-cache capability-probe skip
- `olmlx/engine/loaded_model.py:148` — add `is_image: bool = False`
- `olmlx/engine/inference.py`
  - new `generate_image`
  - `_batch_eligible` (:2485) — add `lm.is_image` to the guard
- `olmlx/app.py` (~:598) — `app.include_router(images.router)`
- `olmlx/config.py` — settings, `audio_max_bytes` naming pattern
- `pyproject.toml` — `[image]` optional extra

### Packaging

`[image]` must be **optional**, alongside `[audio]`. mflux drags in `opencv-python`, `matplotlib`, `fonttools`, `sentencepiece`, `piexif`, `hf-transfer` — roughly 150–250MB. This makes the `ImportError` -> `ValueError` -> 400 path mandatory, not polish.

Add a comment in `pyproject.toml` (matching the existing style) noting that mflux promotes **torch** to an explicit hard requirement of this extra. `CLAUDE.md` currently records that torch is deliberately undeclared because it rides in transitively via xgrammar and mlx-whisper; torch and pillow both sit exactly on mflux's floor, with no headroom below.

Lockfile churn: a fresh resolution floats mlx to 0.32.2 and mlx-vlm to 0.6.17. Both are inside existing pins. Land that as its own commit, separate from the feature.

## Phases

**Phase 1 — load path**
- `registry.py` explicit `"type": "image"` marker + `_detect_model_kind` short-circuit
- `_load_model_image` (resolve declared `hf_path` -> `ModelConfig.from_name`, exact-match guard first) + `is_image` flag and the three guard sites
- `mx.eval(model.parameters())` and `_materialize_module_buffers(model)` on the load thread
- `[image]` extra

**Phase 2 — generation + endpoint**
- `generate_image` under `_inference_locked` + `asyncio.to_thread`
- `routers/images.py` and `schemas/images.py`
- Per-step cancellation via an `in_loop` callback that raises when `cancel_event` is set
- b64_json encoding from `GeneratedImage.image`

**Phase 3 — tests, docs, surfacing**
- `models.json` / `/api/tags` surfacing
- `CLAUDE.md` invariants section
- README

### Per-variant adapter

The two variants are **not** signature-compatible:

| | `QwenImage` (20B) | `QwenImage21` |
|---|---|---|
| LoRA in ctor | `lora_paths`, `lora_scales`, `bake_lora` | none |
| default steps | 4 | 40 |
| default guidance | 4.0 | 1.0 (guidance-free) |
| extra gen args | `pid_decode`, `pid_degrade_sigma` | — |

So `generate_image` needs a thin per-variant dispatch, not one shared call.

## Testing

- **Unit**: registry parsing of `"type": "image"`; `_detect_model_kind` returns `"image"` from the marker **without** attempting a `config.json` fetch; loader dispatch via monkeypatched mflux; router request/response validation; missing-extra -> 400.
- **Misclassification regression (required)**: assert that a plain text entry such as `Qwen/Qwen3-32B-4bit` — which `ModelConfig.from_name` *does* resolve to a Qwen-Image base — is **not** detected as an image model, and that the exact-match alias guard rejects it. This locks in the substring-matcher hazard documented above.
- **Thread-local streams**: a case in `tests/test_thread_local_streams.py` with a negative control, matching the existing `TestScaledRopeBufferMaterialization` / `TestWhisperBufferMaterialization` pattern. Load on one thread, forward on another. This is the class of bug that has bitten us six times (#499, #651, #653, #654, #655, #657, #718) and mflux was developed in a single-threaded CLI where load and forward share a thread, so any latent instance is masked upstream.
- **Guard regression**: `_batch_eligible` returns `False`, prompt-cache probe skipped, adapter base rejected.
- **Live** (`real_model`, Metal-gated): one end-to-end generation at low step count, following `tests/live/test_rerank_real.py`.

## Out of scope for v1

- `response_format: "url"` — no static file server; `b64_json` only
- Image editing (`mflux-generate-qwen-2.1-edit`, reference images, RGBA layers)
- img2img (`image_path` / `image_strength` are accepted by the API but not exposed)
- LoRA
- `n > 1` per request
- Streaming / `partial_images`
- Distributed, speculative, prompt-cache participation — none apply
- Ollama-native surface: Ollama's API has no image-generation endpoint, so there is no parity route to match

## Open questions

1. **Memory co-residency.** ~20–25GB resident for Qwen-Image-2.1 (the ~8B text encoder is held in bf16 — `skip_quantization=True` in the weight definition, because quantizing it degrades semantics — so `-q 8` only shrinks the DiT). Against a 64GB box and `max_loaded_models` sized for LLMs, either it takes a slot and evicts a chat model on every image request, or it gets pinned and permanently costs a third of the budget. This is a policy call, not code.
2. **Should image models count against `max_loaded_models` at all, or get their own slot budget?**
3. **Timeout.** A 40-step 1024x1024 generation will take far longer than any chat request. Does `inference_queue_timeout` need a per-modality override?
4. mflux's own `model.prompt_cache` (text-encoder prefix caching, mflux-community#741) — leave alone in v1, or wire to our lifecycle?

## References

- mflux: https://github.com/mflux-community/mflux
- Qwen-Image-2.1 support: https://github.com/mflux-community/mflux/pull/736
- Reference editing / RGBA / prefix caching: https://github.com/mflux-community/mflux/pull/741
